# Importing Libraries

In [5]:
import pandas as pd
import numpy as np
import os

In [6]:
#creating a shortcut
path = r'C:\Users\16175\Desktop\Instacart Project'

In [7]:
#Shortcut
df = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [10]:
# Create a Dataframe
df_test = pd.DataFrame()

In [11]:
# Create New mixed column
df_test['mix'] = ['a', 'b', 1, True]

In [12]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [13]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [14]:
df_test['mix'] = df_test['mix'].astype('str')

In [17]:
#importing products.csv
df2 = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [19]:
df2.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [20]:
df2_nan = df2[df2['product_name'].isnull() == True]

In [21]:
df2_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [23]:
df2.shape

(49693, 5)

In [24]:
df2_clean = df2[df2['product_name'].isnull() == False]

In [25]:
df2_clean.shape

(49677, 5)

In [26]:
df2_dups = df2_clean[df2_clean.duplicated()]

In [27]:
df2_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [28]:
#Checking df2_clean.shape

(49677, 5)

In [30]:
df2_clean_no_dups = df2_clean.drop_duplicates()

In [32]:
df2_clean_no_dups.shape

(49672, 5)

In [33]:
#exporting data
df2.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task 

In [34]:
df2.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

All columns seem to be numeric- only the prices column should be numeric to give a clear understanding of the values. 
Also, in the prices column, i noticed a clear distance between the 75th percentile and maximum value - 99999 seems to be a mistake for the following reasons
1. its way expensive than most of the other products
2. It appears to be an outlier



In [35]:
#Checking for mixed type data
for col in df.columns.tolist():
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df[weird]) > 0:
    print (col)

There are no mixed-type columns in the orders.csv

In [38]:
# Checking for Missing Values
df.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [37]:
df.shape

(3421083, 7)

For the most part, there are no missing values in all the relevant cells that could be beneficial to my analysis. However, the alarming number on the days_prior columns is worrisome. Although there might be contributing factors
1. New customers
2. Recently purchased 
3. Not recorded

In [40]:
df_nan = df[df['days_since_prior_order'].isnull() == True]

In [47]:
df_nan.shape

(206209, 7)

In [45]:
df.shape

(3421083, 7)

 Missing values appear to be more than 50% of the column and due to this i will use the Median value. The reason for not choosing the mean is due to customer behaviour- some customers prefer to buy in bulk while others prefer weekly or daily shopping, huge prior days might skew my analysis.

In [53]:
df.median()

order_id                  1710542.0
user_id                    102689.0
order_number                   11.0
order_dow                       3.0
order_hour_of_day              13.0
days_since_prior_order          7.0
dtype: float64

In [54]:
df['days_since_prior_order'].fillna(7.0, inplace=True)

In [58]:
df_clean = df['days_since_prior_order'].fillna(7.0, inplace=True)

In [61]:
df_nan = df[df['days_since_prior_order'].isnull() == True]

In [62]:
df_nan

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

In [64]:
df_nan.shape

(0, 7)

In [66]:
#Checkinf for Duplicates

df_dups = df[df.duplicated()]

In [67]:
df_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

In [68]:
df.shape

(3421083, 7)

In [69]:
#exporting data
df.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))